### PR TITLE
Add issuer to teleport-cluster chart

### LINKED ISSUE #10493: Create `Issuer` resource using teleport-cluster chart
### What
I'd like to create `Issuer` resource of cert-manager using [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart.

### How
1. Add `issuer.yaml` to [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart.
1. Add following options to [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart.
    - Option to determine whether or not `Issuer` resource will be create.
    - Option to define `.spec` field of `Issuer` resource.

### Why
Now we must deploy `Issuer` resource manually before deploy [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart. If we can deploy `Issuer` resource with [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart, it is more convenient.

### Workaround
Deploy `Issuer` resource manually before deploy [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart like described at [AWS](https://github.com/gravitational/teleport/blob/master/docs/pages/kubernetes-access/helm/guides/aws.mdx#step-4-install-and-configure-cert-manager) and [GCP](https://github.com/gravitational/teleport/blob/master/docs/pages/kubernetes-access/helm/guides/gcp.mdx#step-4-install-and-configure-cert-manager) guides.

### PR DESCRIPTION
**Pull request by shibataka000**
Close #10493 .

### CODE REVIEW

**Review by zmb3 — CHANGES_REQUESTED** [review]
I see this changed in docs and examples, but I don't see any templates using the new issuer spec. Am I missing something?

**Comment by shibataka000** [comment]
@zmb3 Could you review this PR please? Thank you!

**Comment by shibataka000** [comment]
@ptgott @r0mant @xinding33 @zmb3 Could you review this PR please? Thank you!

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 1]
```diff
@@ -662,6 +662,60 @@ Setting `highAvailability.certManager.addCommonName` to `true` will instruct `ce
...
+  --set highAvailability.certManager.issuerSpec.acme.email=email@address.com \
+  --set highAvailability.certManager.issuerSpec.acme.server=https://acme-v02.api.letsencrypt.org/directory \
+  --set highAvailability.certManager.issuerSpec.acme.privateKeySecretRef.name=letsencrypt-production \
+  --set highAvailability.certManager.issuerSpec.acme.solvers[0].selector.dnsZones[0]=example.com \
+  --set highAvailability.certManager.issuerSpec.acme.solvers[0].dns01.route53.region=us-east-1 \
+  --set highAvailability.certManager.issuerSpec.acme.solvers[0].dns01.route53.hostedZoneID=Z0159221358P96JYAUAA4
+  ```
+  </TabItem>
+</Tabs>
+
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:718]: I think this definition being repeated twice in adjacent sections of the docs is probably redundant.

If you always need to provide your own `issuerSpec` when `createIssuer: true`, maybe the only value added should be `issuerSpec`? That way, the chart could automatically skip creation when `issuerSpec: {}`, but carry out creation when `issuerSpec` evaluates to a non-empty value.

```suggestion
```

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 2]
```diff
@@ -741,6 +795,62 @@ Sets the `Group` of `Issuer` to be used when issuing certificates with `cert-man
...
+Spec of `Issuer` or `ClusterIssuer`.
+
+<Admonition type="note" title="Installing cert-manager">
+  You must install and configure `cert-manager` in your Kubernetes cluster yourself.
+
+  See the [cert-manager Helm install instructions](https://cert-manager.io/docs/installation/kubernetes/#option-2-install-crds-as-part-of-the-helm-release)
+  and the relevant sections of the [AWS](../guides/aws.mdx#step-47-install-and-configure-cert-manager) and [GCP](../guides/gcp.mdx#step-47-install-and-configure-cert-manager) guides for more information.
+
+  You must create this yourself when `highAvailability.certManager.createIssuer` is `true`.
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:812]: See comments above about removing `createIssuer`.

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 3]
```diff
@@ -741,6 +795,62 @@ Sets the `Group` of `Issuer` to be used when issuing certificates with `cert-man
   </TabItem>
 </Tabs>
 
+### `highAvailability.certManager.issuerSpec`
+
+| Type | Default value | Can be used in `custom` mode? | `teleport.yaml` equivalent |
+| - | - | - | - |
+| `object` | `{}` | ❌ | None |
+
+Spec of `Issuer` or `ClusterIssuer`.
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:804]: I don't see a way to toggle from `Issuer` to `ClusterIssuer` currently?

**Review by webvictim — CHANGES_REQUESTED** [review]
Hey @shibataka000 - thanks for the PR. Could you please take a look at our Helm chart contribution guidelines? https://github.com/gravitational/teleport/blob/master/examples/chart/CONTRIBUTING.md

We'd need some unit tests adding for this new functionality.

Overall I'm honestly still a little on the fence about this PR - it feels like we're moving something that should be outside the chart inside the chart. Could you describe why having the `Issuer` created by this chart would be preferable to creating it outside of Helm? An `Issuer` is generally something that's provisioned outside of a release once, and then used by multiple releases. This feels like we're moving the lifecycle to a place it shouldn't be.

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx (hunk 3 above)
Comment (shibataka000) [reply docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx]: You can do it by [`highAvailability.certManager.issuerKind`](https://goteleport.com/docs/kubernetes-access/helm/reference/teleport-cluster/#highavailabilitycertmanagerissuerkind) option.

Comment (shibataka000) [reply docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx]: You can do it by [`highAvailability.certManager.issuerKind`](https://goteleport.com/docs/kubernetes-access/helm/reference/teleport-cluster/#highavailabilitycertmanagerissuerkind) option.

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx (hunk 2 above)
Comment (shibataka000) [reply docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx]: I fixed it by 290e52289394097dcc2d30cd3ec8357528e491c6 .

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx (hunk 1 above)
Comment (shibataka000) [reply docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx]: I fixed it by 290e52289394097dcc2d30cd3ec8357528e491c6 .

**Comment by shibataka000** [comment]
@webvictim Thank you for your review!

> We'd need some unit tests adding for this new functionality.

I added some unit tests by b8ba6bf2f2692c38de3d0e9853ed923f0bacce4c .

> Could you describe why having the `Issuer` created by this chart would be preferable to creating it outside of Helm?

Because I think it is more convenient if  [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart has `Issuer` resource.

We need 3 steps to install teleport using helm charts now.

1. Deploy [cert-manager](https://artifacthub.io/packages/helm/cert-manager/cert-manager) chart.
2. Deploy `Issue` resource.
3. Deploy [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart.

If [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart have `Issuer` resource, we can install teleport by only 2 steps.

1. Deploy [cert-manager](https://artifacthub.io/packages/helm/cert-manager/cert-manager) chart. `Issue` resource is also deployed.
2. Deploy [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart.

I described above in #10493 .

For example, [AWS Load Balancer Controller's helm chart](https://github.com/aws/eks-charts/tree/master/stable/aws-load-balancer-controller) has [`Issuer`](https://github.com/aws/eks-charts/blob/2f4723805dd92363f1b042add61df688392fd560/stable/aws-load-balancer-controller/templates/webhook.yaml#L164-L177) resource. We can install AWS Load Balancer Controller and `Issuer` resource by only 1 step. I think it is convenient.

Deploying `Issuer` resource with [teleport-cluster](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster) chart is optional. If lifecycle of teleport and `Issuer` are difference in your environment, you can set `highAvailability.certManager.issuerSpec={}` and deploy `Issuer` manually.

**Comment by hugoShaka** [comment]
Closing because the related feature-request has been declined.

Please see the feature-request issue for more details: https://github.com/gravitational/teleport/issues/10493#issuecomment-1239727435

**Pull request closed without merging**